crates/backend/src/codegen.rs

```diff
@@ -971,22 +971,42 @@ impl TryToTokens for ast::ImportFunction {
                 );
             }
             Some(ref ty) => {
-                abi_ret = quote! {
-                    <#ty as wasm_bindgen::convert::FromWasmAbi>::Abi
-                };
-                convert_ret = quote! {
-                    <#ty as wasm_bindgen::convert::FromWasmAbi>
-                        ::from_abi(#ret_ident)
-                };
+                if self.function.r#async {
+                    abi_ret = quote! { js_sys::Promise };
+                    let future = quote! { wasm_bindgen_futures::JsFuture::from(#ret_ident).await };
+                    convert_ret = if self.catch {
+                        quote! { Ok(<#ty as wasm_bindgen::FromJsValue>::from_js_value(#future?)) }
```

Comment:
Could this perhaps use `TryFrom` instead of a new trait? (now that I think about it).

It might be cool if this could leverage existing `TryFrom` implementation (and add a few here too), and that way if the conversion from `JsValue` fails we've got a way to return the error (it's like as if the future failed)

Reply:
Actually I tried that but it didn't go well. I wanted these blanket impls:
```
impl<T: JsCast> TryFrom<JsValue> for T { ... }
impl<T: TryFrom<JsValue>> TryFrom<JsValue> for Option<T> { ... }
```
But these resulted in a lot of conflicting implementations.

Maybe this trait could be leveraged into `JsCast`, but a new trait looked cleaner to me.

Reply:
Oh I don't think we'll be able to do some blanket impls, but we could add specific impls perhaps?

In general I'm just hoping that we can propagate an error for a failed conversion.

In general though this seems like a really tricky problem. If you'd prefer we can just stick to `Result<JsValue, JsValue>` like `Promise` and call it a day.

Reply:
Actually, returning `Result<JsValue, JsValue>` was my first option, but it looked _poor_. The user can just `map` to do the casting, but it breaks ergonomics.

And at first I though that I could coerce the error in the conversion to that `Err(_)` but it did not work either. On one side, because a cach function is not required to return a `JsValue` but any type that implements `From<JsValue>`, because of the `?` operator. But a cast failure would be a `js_sys::TypeError`, and that conversion may not be possible.
On the other side, `wasm_bindgen` does not actually depend on `js_sys`. Adding that dependency just for this seemed overkill.

The fancy solution could be making `JsFuture` generic and doing ABI magic in the JS side, just like regular functions do. But that is way out of my league.

Anyway, I'm not convinced with the _failing conversions_ argument. If I write a regular JS function and import it with a wrong prototype it will just panic or return a default value, depending on the types, (I think these situations are not documented):

```javascript
function foo() { return 42; }
```
```rust
#[wasm_bindgen]
extern "C" {
    fn foo() -> String;  //undefined?
}
```
Going async would be quite similar:
```javascript
async function foo() { return 42; }
```
```rust
#[wasm_bindgen]
extern "C" {
    async fn foo() -> String; //undefined!
}
```